### ISSUE TITLE
Using optax with partially complex models

### ISSUE THREAD

**Issue by GeoffNN** [google-deepmind/optax#287]
Hello!

I'm trying to implement the Fourier Neural Operator model from https://github.com/devzhk/PINO/, translating the Pytorch code there into jax.
I ran into an issue where the optax optimizer_state becomes complex for _all_ weight matrices of my model, rather than only the leaves corresponding to complex weights in the model. This only happens after a while, and I've only been able to reproduce this while using an optax scheduler. Using only `optax.adam` seems to work without issue.

Full reproducing script: https://colab.research.google.com/drive/1a5BFz0G20t_VvXOfQqJfqHo4-HL3Ok3X?usp=sharing.

EDIT: This is related to (but possibly different than) https://github.com/deepmind/optax/issues/196

**Comment by PhilipVinc** [comment]
By the way, can you try using master? The current release of optax does not include the changes to work with complex numbers.

Current master silently computes the wrong norm in adam.

**Comment by GeoffNN** [comment]
Is that true even when passing the conjugated gradients to `opt.update` ?

**Comment by GeoffNN** [comment]
I understood your earlier comment on Adam. It's currently https://github.com/deepmind/optax/blob/5fc395b5e49602e3c0027b976d1368755b20cd0b/optax/_src/transform.py#L78, but for 2nd moment, should have the complex module ^2, as in https://github.com/pytorch/pytorch/blob/3a03af2f505494582f1359254431f0c47fd88705/torch/optim/_functional.py#L98

**Comment by GeoffNN** [comment]
I think my initial observation was due to the gradient clipping. The code for computing the global norm for clipping has the same issue as Adam, and squares a potentially complex number, rather than taking the squared module.
https://github.com/deepmind/optax/blob/5fc395b5e49602e3c0027b976d1368755b20cd0b/optax/_src/clipping.py#L102

**Comment by PhilipVinc** [comment]
> optax should at least output some form of warning when handling complex parameters...

Yeah, That's what we tried to argue with optax developers. Hopefully once #297 is merged this will be addressed.

> I think my initial observation was due to the gradient clipping. The code for computing the global norm for clipping has the same issue as Adam, and squares a potentially complex number, rather than taking the squared module.

@wdphy16 does PR #279  address gradient clipping as well?

**Comment by wdphy16** [comment]
Yes, we're working on this. I just tried your code that if you install my branch:
```
!pip install git+git://github.com/wdphy16/optax.git@complex_optimizer
```
the code runs without error. I hope that PR can get merged soon.

**Comment by GeoffNN** [comment]
Ah thanks! I have a local messy fix, I'll migrate it to yours.

Also, I was pointed to this issue in jax, which you might weigh in on. https://github.com/google/jax/issues/4891
The question is whether jax.grad should return the already conjugated gradient or not.

**Comment by mtthss** [comment]
Did PR #279 address your issue? Anything else you want to discuss or can I close the issue?
